src/QueryAggregator.js

```diff
@@ -0,0 +1,109 @@
+import invariant from 'invariant';
+import Relay from 'react-relay';
+
+import getParamsForRoute from './getParamsForRoute';
+
+export default class QueryAggregator {
+  constructor() {
+    this.uniqueQueryNames = new WeakMap();
+    this.lastQueryIndex = 0;
+
+    this.data = {};
+  }
+
+  updateRoute(routerProps) {
+    const {params, location} = routerProps;
+
+    const relayRoute = {
+      queries: {},
+      params: {},
+    };
+    const fragmentSpecs = {};
+
+    routerProps.branch.forEach(route => {
+      const {queries} = route;
+      if (!queries) {
+        return;
+      }
+
+      const {component} = route;
+
+      // In principle not all container component routes have to specify
+      // queries, because some of them might somehow receive fragments from
+      // their parents, but it would definitely be wrong to specify queries
+      // for a component that isn't a container.
+      invariant(
+        Relay.isContainer(component),
+        'relay-nested-routes: Route with queries specifies component `%s` ' +
+        'that is not a Relay container.',
+        component.displayName || component.name
+      );
+
+      const routeParams = getParamsForRoute({route, params, location});
+      relayRoute.params = Object.assign(relayRoute.params, routeParams);
```

Comment:
That's slightly different - I'm intentionally updating `relayRoute.params` in-place because... well, why not? Nothing else is using it.